Mathlib/Geometry/Manifold/ContMDiff/Basic.lean

```diff
@@ -259,6 +259,55 @@ alias contMDiffAt_of_not_mem := contMDiffAt_of_notMem
 alias contMDiffAt_of_not_mem_mulTSupport := contMDiffAt_of_notMem_mulTSupport
 
 
+/-! ### Being `C^k` on a set can be tested in its union -/
+section contMDiff_union
+
+variable {s t : Set M}
+
+/-- If a function is `C^k` on two open sets, it is also `C^n` on their union. -/
+lemma ContMDiffOn.union_of_isOpen (hf : ContMDiffOn I I' n f s) (hf' : ContMDiffOn I I' n f t)
+    (hs : IsOpen s) (ht : IsOpen t) :
+    ContMDiffOn I I' n f (s ∪ t) := by
+  intro x hx
+  obtain (hx | hx) := hx
+  · exact (hf x hx).contMDiffAt (hs.mem_nhds hx) |>.contMDiffWithinAt
+  · exact (hf' x hx).contMDiffAt (ht.mem_nhds hx) |>.contMDiffWithinAt
+
+/-- A function is `C^k` on two open sets iff it is `C^k` on their union. -/
+lemma contMDiffOn_union_iff_of_isOpen (hs : IsOpen s) (ht : IsOpen t) :
+    ContMDiffOn I I' n f (s ∪ t) ↔ ContMDiffOn I I' n f s ∧ ContMDiffOn I I' n f t :=
+  ⟨fun h ↦ ⟨h.mono subset_union_left, h.mono subset_union_right⟩,
+   fun ⟨hfs, hft⟩ ↦ ContMDiffOn.union_of_isOpen hfs hft hs ht⟩
+
+lemma contMDiff_of_contMDiffOn_union_of_isOpen (hf : ContMDiffOn I I' n f s)
+    (hf' : ContMDiffOn I I' n f t) (hst : s ∪ t = univ) (hs : IsOpen s) (ht : IsOpen t) :
+    ContMDiff I I' n f := by
+  rw [← contMDiffOn_univ, ← hst]
+  exact hf.union_of_isOpen hf' hs ht
+
+/-- If a function is `C^k` on open sets `s i`, it is `C^k` on their union -/
+lemma ContMDiffOn.iUnion_of_isOpen {ι : Type*} {s : ι → Set M}
+    (hf : ∀ i : ι, ContMDiffOn I I' n f (s i)) (hs : ∀ i, IsOpen (s i)) :
+    ContMDiffOn I I' n f (⋃ i, s i) := by
+  rintro x ⟨si, ⟨i, rfl⟩, hxsi⟩
+  exact (hf i).contMDiffAt ((hs i).mem_nhds hxsi) |>.contMDiffWithinAt
+
+/-- A function is `C^k` on a union of open sets `s i` iff it is `C^k` on each `s i`. -/
+lemma contMDiffOn_iUnion_iff_of_isOpen  {ι : Type*} {s : ι → Set M}
```

Comment:
```suggestion
lemma contMDiffOn_iUnion_iff_of_isOpen {ι : Type*} {s : ι → Set M}
```
(I thought we had a linter for that now)

Reply:
We should have, this is unexpected. CC @adomani

Reply:
Not really: the parent of this commit is `7672327e9c8b73aeb0930dcf8389e1dc0a1a729a`, which landed in mathlib today.

Reply:
:scream: `lemma` has kind `lemma`, not declaration!! So, the linter ignores lemmas... :facepalm:

Reply:
#26706, although I imagine that there will be some fixing to do, before it can be merged.

Reply:
If you are willing to help with fixing spaces, I'd be very happy to review and merge the PRs!